### ISSUE TITLE
Raft-based ops: getting rid of the old stuff

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#15422]
Once we finish the topology-on-raft project, which mostly consists of adding code, we should also start seriously considering removing some code.

In particular, it would be nice if in the future we could remove the legacy code, like gossiper-based schema and topology operations.

One major obstacle that prevents us from doing that is RECOVERY mode. If we lose group 0 majority, we can use RECOVERY to revert the gossiper-based way, remove dead nodes and fix schema using gossiper-based ops, and then perform upgrade-to-raft procedure to establish a new group 0.

This means that as long as we have RECOVERY mode as it is now, we will never be able to get rid of this old code.

One way to approach the problem would be to implement a different kind of RECOVERY where instead of using gossiper for topology ops, we use an external orchestration tool to make centralized decisions about cluster topology  and schema without group 0 majority being able to.

Original discussion:
https://github.com/scylladb/scylladb/pull/15331#discussion_r1321923707

cc @avikivity @gleb-cloudius

**Comment by kbr-scylla** [comment]
Estimating as M, but that doesn't take into account the pre-requisite https://github.com/scylladb/scylladb/issues/18791, only the "cleaning up" part.

**Comment by kbr-scylla** [comment]
Blocked by https://github.com/scylladb/scylladb/issues/18791

**Comment by kbr-scylla** [comment]
Optimistically planning for Q1/25, but currently blocked by https://github.com/scylladb/scylladb/issues/20657 which I hope will be finished soon.

cc @patjed41 @emdroid

**Comment by patjed41** [comment]
Another blocker for removing the old stuff is making the Raft-based topology mandatory. So, the plan should be as follows:
1. Finish #20657.
2. Make the Raft-based topology mandatory.
3. Remove the old stuff, which includes:
    - the gossip-based topology with all the code that is specific to it (like schema, CDC),
    - all "upgrade to Raft" procedures, like upgrade-to-raft and upgrade-to-raft-topology,
    - the old recovery procedure (tests, documentation).

We may also remove the RECOVERY mode but Maintenance Mode uses it, so I'm not sure what we will do. Maybe we will just rename it to MAINTENANCE.

We will (hopefully) backport #20657 to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.

**Comment by mykaul** [comment]
> We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.

Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.

**Comment by patjed41** [comment]
> > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> 
> Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.

Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.

**Comment by mykaul** [comment]
> > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > 
> > 
> > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> 
> Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.

It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.

**Comment by gleb-cloudius** [comment]
> > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > 
> > > 
> > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > 
> > 
> > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> 
> It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.

Why 2026.1 and not 2025.3?

**Comment by mykaul** [comment]
> > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > 
> > > > 
> > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > 
> > > 
> > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > 
> > 
> > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> 
> Why 2026.1 and not 2025.3?

2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.

**Comment by gleb-cloudius** [comment]
> > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > 
> > > > > 
> > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > 
> > > > 
> > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > 
> > > 
> > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > 
> > 
> > Why 2026.1 and not 2025.3?
> 
> 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.

So lets do it now for 2025.4. What is the previous LTS and the next one?

**Comment by mykaul** [comment]
> > > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > > 
> > > > > > 
> > > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > > 
> > > > > 
> > > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > > 
> > > > 
> > > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > > 
> > > 
> > > Why 2026.1 and not 2025.3?
> > 
> > 
> > 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.
> 
> So lets do it now for 2025.4. What is the previous LTS and the next one?

Previous LTS  - 2024.1
Current LTS - 2025.1
Next LTS - 2026.1.

**Comment by gleb-cloudius** [comment]
> > > > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > > > 
> > > > > > > 
> > > > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > > > 
> > > > > > 
> > > > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > > > 
> > > > > 
> > > > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > > > 
> > > > 
> > > > Why 2026.1 and not 2025.3?
> > > 
> > > 
> > > 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.
> > 
> > 
> > So lets do it now for 2025.4. What is the previous LTS and the next one?
> 
> Previous LTS - 2024.1 Current LTS - 2025.1 Next LTS - 2026.1.

So I am confused. What prevented us from making it mandatory now? @patjed41 was recovery procedure backported to 2025.1?

BTW https://docs.scylladb.com/manual/stable/troubleshooting/handling-node-failures.html says that raft topology is mandatory in 2025.2.

**Comment by mykaul** [comment]
> > > > > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > > > > 
> > > > > > > 
> > > > > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > > > > 
> > > > > > 
> > > > > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > > > > 
> > > > > 
> > > > > Why 2026.1 and not 2025.3?
> > > > 
> > > > 
> > > > 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.
> > > 
> > > 
> > > So lets do it now for 2025.4. What is the previous LTS and the next one?
> > 
> > 
> > Previous LTS - 2024.1 Current LTS - 2025.1 Next LTS - 2026.1.
> 
> So I am confused. What prevented us from making it mandatory now? [@patjed41](https://github.com/patjed41) was recovery procedure backported to 2025.1?

I honestly don't know. I had hoped so (https://github.com/scylladb/scylladb/issues/15422#issuecomment-2595751929 ).
Maybe I mis-remember some detail on why we could not do it.

**Comment by gleb-cloudius** [comment]
> > > > > > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > > > > > 
> > > > > > > 
> > > > > > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > > > > > 
> > > > > > 
> > > > > > Why 2026.1 and not 2025.3?
> > > > > 
> > > > > 
> > > > > 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.
> > > > 
> > > > 
> > > > So lets do it now for 2025.4. What is the previous LTS and the next one?
> > > 
> > > 
> > > Previous LTS - 2024.1 Current LTS - 2025.1 Next LTS - 2026.1.
> > 
> > 
> > So I am confused. What prevented us from making it mandatory now? [@patjed41](https://github.com/patjed41) was recovery procedure backported to 2025.1?
> 
> I honestly don't know. I had hoped so ([#15422 (comment)](https://github.com/scylladb/scylladb/issues/15422#issuecomment-2595751929) ). Maybe I mis-remember some detail on why we could not do it.

@avikivity can you clarify. IIRC you killed the idea to make raft topology mandatory in 2025.2, but nobody remember why.

**Comment by patjed41** [comment]
> So I am confused. What prevented us from making it mandatory now? @patjed41 was recovery procedure backported to 2025.1?

No, we considered it, but in the end I didn't do it. It turned out that backporting and maintaining it would require a lot of effort. I also don't think this feature is critical for 2025.1.

> BTW https://docs.scylladb.com/manual/stable/troubleshooting/handling-node-failures.html says that raft topology is mandatory in 2025.2.

The documentation assumes that the Raft-based topology is mandatory from 2025.2. The reason is that it makes the documentation simpler. We don't have to have both recovery procedures documented, we could remove handling failures in the gossip-based topology, etc. My recent PR made this change: #24583. The idea was from https://github.com/scylladb/scylladb/issues/24348#issuecomment-2955755287.

Upgrade guides to 2025.1 and to 2025.2 state that enabling the Raft-based topology is a mandatory step. If a user still somehow doesn't use it in 2025.2, we redirect them to the relevant documentation in older versions (upgrade guides containing the upgrade-to-raft-topology procedure, the old recovery procedure).

We still have to make the Raft-based topology mandatory in the code.

**Comment by gleb-cloudius** [comment]
> > 
> > I honestly don't know. I had hoped so ([#15422 (comment)](https://github.com/scylladb/scylladb/issues/15422#issuecomment-2595751929) ). Maybe I mis-remember some detail on why we could not do it.
> 
> [@avikivity](https://github.com/avikivity) can you clarify. IIRC you killed the idea to make raft topology mandatory in 2025.2, but nobody remember why.

@avikivity ping

**Comment by avikivity** [comment]
> > > > > > > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > > > > > > 
> > > > > > > > > > 
> > > > > > > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > > > > > > 
> > > > > > > 
> > > > > > > Why 2026.1 and not 2025.3?
> > > > > > 
> > > > > > 
> > > > > > 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.
> > > > > 
> > > > > 
> > > > > So lets do it now for 2025.4. What is the previous LTS and the next one?
> > > > 
> > > > 
> > > > Previous LTS - 2024.1 Current LTS - 2025.1 Next LTS - 2026.1.
> > > 
> > > 
> > > So I am confused. What prevented us from making it mandatory now? [@patjed41](https://github.com/patjed41) was recovery procedure backported to 2025.1?
> > 
> > 
> > I honestly don't know. I had hoped so ([#15422 (comment)](https://github.com/scylladb/scylladb/issues/15422#issuecomment-2595751929) ). Maybe I mis-remember some detail on why we could not do it.
> 
> [@avikivity](https://github.com/avikivity) can you clarify. IIRC you killed the idea to make raft topology mandatory in 2025.2, but nobody remember why.

Replaying the same inputs through the same evaluation function, probably the ability to upgrade.

2025.1 is the first LTS to support Raft topology. If we remove gossip topology in 2025.1.z (z > 1), then we remove the ability yo upgrade from 2024.1 to 2026.1 and instead require 2024.1 -> 2025.1 -> 2026.1.

**Comment by gleb-cloudius** [comment]
> > > > > > > > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > > > > > > > 
> > > > > > > > > > > 
> > > > > > > > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > > > > > > > 
> > > > > > > > > > 
> > > > > > > > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Why 2026.1 and not 2025.3?
> > > > > > > 
> > > > > > > 
> > > > > > > 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.
> > > > > > 
> > > > > > 
> > > > > > So lets do it now for 2025.4. What is the previous LTS and the next one?
> > > > > 
> > > > > 
> > > > > Previous LTS - 2024.1 Current LTS - 2025.1 Next LTS - 2026.1.
> > > > 
> > > > 
> > > > So I am confused. What prevented us from making it mandatory now? [@patjed41](https://github.com/patjed41) was recovery procedure backported to 2025.1?
> > > 
> > > 
> > > I honestly don't know. I had hoped so ([#15422 (comment)](https://github.com/scylladb/scylladb/issues/15422#issuecomment-2595751929) ). Maybe I mis-remember some detail on why we could not do it.
> > 
> > 
> > [@avikivity](https://github.com/avikivity) can you clarify. IIRC you killed the idea to make raft topology mandatory in 2025.2, but nobody remember why.
> 
> Replaying the same inputs through the same evaluation function, probably the ability to upgrade.
> 
> 2025.1 is the first LTS to support Raft topology. If we remove gossip topology in 2025.1.z (z > 1), then we remove the ability

You mean `s/2025.1.z/2025.z` right?

> yo upgrade from 2024.1 to 2026.1 and instead require 2024.1 -> 2025.1 -> 2026.1.

And is this a problem? Because if we want to support direct upgrade from 2024.1  to 2026.1 it means that we cannot remove gossiper topology from 2026.1 as well.

**Comment by avikivity** [comment]
> > > > > > > > > > > > > We will (hopefully) backport [#20657](https://github.com/scylladb/scylladb/issues/20657) to 2025.1. and if not, it will be in 2025.2. I think we should make the Raft-based topology mandatory in 2025.2 (it shouldn't require much work) even if we don't remove the old stuff in 2025.2. We will at least be able to stop testing the gossip-based topology, which is a huge gain, especially for CI.
> > > > > > > > > > > > 
> > > > > > > > > > > > 
> > > > > > > > > > > > Yes, Raft based topology has to be mandatory in 2025.2, that's the plan.
> > > > > > > > > > > 
> > > > > > > > > > > 
> > > > > > > > > > > Did we make Raft-based topology mandatory in 2025.2 in the end? I can't find a PR or anything else related.
> > > > > > > > > > 
> > > > > > > > > > 
> > > > > > > > > > It's enabled by default, but you can upgrade from previous versions and still NOT use Raft. Should be done for 2026.1.
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > Why 2026.1 and not 2025.3?
> > > > > > > > 
> > > > > > > > 
> > > > > > > > 2025.3 was already branched, so it's a bit late for it. It could be in 2025.4. Generally, it should be done before the next LTS, as we do support LTS to LTS upgrades.
> > > > > > > 
> > > > > > > 
> > > > > > > So lets do it now for 2025.4. What is the previous LTS and the next one?
> > > > > > 
> > > > > > 
> > > > > > Previous LTS - 2024.1 Current LTS - 2025.1 Next LTS - 2026.1.
> > > > > 
> > > > > 
> > > > > So I am confused. What prevented us from making it mandatory now? [@patjed41](https://github.com/patjed41) was recovery procedure backported to 2025.1?
> > > > 
> > > > 
> > > > I honestly don't know. I had hoped so ([#15422 (comment)](https://github.com/scylladb/scylladb/issues/15422#issuecomment-2595751929) ). Maybe I mis-remember some detail on why we could not do it.
> > > 
> > > 
> > > [@avikivity](https://github.com/avikivity) can you clarify. IIRC you killed the idea to make raft topology mandatory in 2025.2, but nobody remember why.
> > 
> > 
> > Replaying the same inputs through the same evaluation function, probably the ability to upgrade.
> > 2025.1 is the first LTS to support Raft topology. If we remove gossip topology in 2025.1.z (z > 1), then we remove the ability
> 
> You mean `s/2025.1.z/2025.z` right?

Right. What's the point of spelling everything out carefully, if it's wrong?

> 
> > yo upgrade from 2024.1 to 2026.1 and instead require 2024.1 -> 2025.1 -> 2026.1.
> 
> And is this a problem?

Yes. Customers try to avoid the upgrade treadmill.

>  Because if we want to support direct upgrade from 2024.1 to 2026.1 it means that we cannot remove gossiper topology from 2026.1 as well.

Possibly, I hate it too.

**Comment by gleb-cloudius** [comment]
> > > yo upgrade from 2024.1 to 2026.1 and instead require 2024.1 -> 2025.1 -> 2026.1.
> > 
> > 
> > And is this a problem?
> 
> Yes. Customers try to avoid the upgrade treadmill.
> 

I hate upgrades probably more than average person (and hence upgrade Fedora only after my current version stop been supported). But upgrading once a year seams reasonable.

**Comment by avikivity** [comment]
> > > > yo upgrade from 2024.1 to 2026.1 and instead require 2024.1 -> 2025.1 -> 2026.1.
> > > 
> > > 
> > > And is this a problem?
> > 
> > 
> > Yes. Customers try to avoid the upgrade treadmill.
> 
> I hate upgrades probably more than average person (and hence upgrade Fedora only after my current version stop been supported). But upgrading once a year seams reasonable.

Database operators are even more conservative.

**Comment by mykaul** [comment]
@emdroid - is this on track to 2026.2?

**Comment by gleb-cloudius** [comment]
> [@emdroid](https://github.com/emdroid) - is this on track to 2026.2?

I am working on the patch to deprecate the option and @scylladb/releng need to drop gossiper topology job from CI and other runs.
